main.rb

```diff
@@ -1,26 +1,266 @@
 module Enumerable
   def my_each
-    if self.is_a? Range then
+    if is_a? Range
       min = self.min
       max = self.max
-      
-      while min < max do
+
+      while min <= max
         puts min
         min += 1
-      
       end
-    else
+
+    elsif block_given?
       min = 0
       max = length
+
       while min < max
         yield(self[min])
         min += 1
       end
+    else
+      to_enum(:my_each)
+    end
+  end
+
+  def my_each_with_index
+    if is_a? Range
+      min = self.min
+      max = self.max
+      index = 0
+      while min <= max
+        yield(min, index)
+        min += 1
+        index += 1
+      end
+
+    elsif block_given?
+      min = 0
+      max = length
+
+      while min < max
+        yield(self[min], min)
+        min += 1
+      end
+    else
+      to_enum(:my_each_with_index)
+    end
+  end
+
+  def my_select
+    if block_given?
+      items = []
+      my_each do |item|
+        items << item if yield(item)
+      end
+      items
+    else
+      to_enum(:my_select)
+    end
+  end
+
+  def my_all?
+    bool_item = true
+    if block_given?
+      my_each do |item|
+        bool_item = false unless yield(item)
+      end
     end
+    bool_item
   end
+
+  def my_any?
+    bool_item = true
+    if block_given?
+      bool_item = false
+      my_each do |item|
+        bool_item = true if yield(item)
+      end
+    end
+    bool_item
+  end
+end
+
+module Enumerable
+  def my_none?
+    if block_given?
+      bool_item = true
+      my_each do |item|
+        bool_item = false if yield(item)
+      end
+    else
+      bool_item = false
+    end
+    bool_item
+  end
+
+  def my_count
+    if block_given?
+      counter = 0
+      my_each do |item|
+        counter += 1 if yield(item)
+      end
+    else
+      counter = length
+    end
+    counter
+  end
+
+  def my_map(is_proc = nil)
+    if block_given?
+      new_items = []
+      my_each do |item|
+        is_proc ? new_items.push(is_proc.call(item)) : new_items << yield(item)
+      end
+      new_items
+    else
+      to_enum(:my_map)
+    end
+  end
+
+  def my_inject(temp_option = nil)
```

Comment:
* [ ] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```p (5..10).my_inject { |sum, n| sum + n } != p (5..10).inject { |sum, n| sum + n }```
```p (5..10).my_inject(1) { |product, n| product * n } != p (5..10).inject(1) { |product, n| product * n }```